src/libprojectM/Renderer/MilkdropWaveform.cpp

```diff
@@ -4,393 +4,613 @@
-        mat_first_translation[3][1] = -0.5;
-
-        glm::mat4 mat_scale = glm::mat4(1.0);
-        mat_scale[0][0] = aspectScale;
-
-        float s = glm::sin(glm::radians(-rot));
-        float c = glm::cos(glm::radians(-rot));
-        glm::mat4 mat_rotation = glm::mat4(c, -s, 0, 0,
-                                           s, c, 0, 0,
-                                           0, 0, 1, 0,
-                                           0, 0, 0, 1);
-
-        glm::mat4 mat_second_translation = glm::mat4(1.0);
-        mat_second_translation[3][0] = 0.5;
-        mat_second_translation[3][1] = 0.5;
-
-        glm::mat4 mat_vertex = context.mat_ortho;
-        mat_vertex = mat_first_translation * mat_vertex;
-        mat_vertex = mat_scale * mat_vertex;
-        mat_vertex = mat_rotation * mat_vertex;
-        mat_vertex = mat_second_translation * mat_vertex;
-        glUniformMatrix4fv(context.uniform_v2f_c4f_vertex_tranformation, 1, GL_FALSE, glm::value_ptr(mat_vertex));
-
-        if (modulateAlphaByVolume) ModulateOpacityByVolume(context);
-        else temp_a = a;
+        temp_a = a;
+        if (modulateAlphaByVolume)
+        {
+            ModulateOpacityByVolume(context);
+        }
         MaximizeColors(context);
 
-#ifndef GL_TRANSITION
-        if (dots == 1) glEnable(GL_LINE_STIPPLE);
-#endif
+        // Additive wave drawing (vice overwrite)
+        if (additive == 1)
+        {
+            glBlendFunc(GL_SRC_ALPHA, GL_ONE);
+        }
+        else
+        {
+            glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+        }
+
+        // Always draw "thick" dots.
+        auto iterations = thick || dots ? 4 : 1;
 
-        //Thick wave drawing
-        if (thick == 1) glLineWidth((context.texsize < 512) ? 2 : 2 * context.texsize / 512);
-        else glLineWidth((context.texsize < 512) ? 1 : context.texsize / 512);
+        // Need to use +/- 1.0 here instead of 2.0 used in Milkdrop to achieve the same rendering result.
+        auto incrementX = 1.0f / static_cast<float>(context.viewportSizeX);
+        auto incrementY = 1.0f / static_cast<float>(context.viewportSizeY);
 
-        //Additive wave drawing (vice overwrite)
-        if (additive == 1)glBlendFunc(GL_SRC_ALPHA, GL_ONE);
-        else glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+        GLuint drawType = dots ? GL_POINTS : (loop ? GL_LINE_LOOP : GL_LINE_STRIP);
 
         glBindVertexArray(m_vaoID);
+        glBindBuffer(GL_ARRAY_BUFFER, m_vboID);
 
-        if (loop)
-            glDrawArrays(GL_LINE_LOOP, 0, samples);
-        else
-            glDrawArrays(GL_LINE_STRIP, 0, samples);
+        // If thick outline is used, draw the shape four times with slight offsets
+        // (top left, top right, bottom right, bottom left).
+        for (auto iteration = 0; iteration < iterations; iteration++)
+        {
+            switch (iteration)
+            {
+                case 0:
+                default:
```

Comment:
It looks so very wrong to put the `default` at the top...
Do we expect it to always have four iterations? maybe it should blow up if more?

Reply:
The `iterations` value is either 1 or 4, this is hard-coded a few lines above. I'll change it to a const int to make it even clearer, and just drop the default, as the compiler will know about the possible values and not emit a note or warning.